src/Logic.hs

```diff
@@ -459,12 +464,14 @@ pushCandidate (pullRequestId, pullRequest) state = do
   -- the pull request has really been approved and built successfully. If it was
   -- not, there is a bug in the program.
   let approved  = isJust $ Pr.approvedBy pullRequest
-      succeeded = Pr.buildStatus pullRequest == BuildSucceeded
       status    = Pr.integrationStatus pullRequest
       prBranch  = Pr.branch pullRequest
-      newHead   = assert (approved && succeeded) $ case status of
-        Integrated sha -> sha
-        _              -> error "inconsistent state: build succeeded for non-integrated pull request"
+      newHead   = assert approved $ case status of
+        Integrated sha BuildSucceeded -> sha
+        Integrated _ _notSucceeded ->
+          error "Trying to push a candidate for which the build did not pass."
+        _notIntegrated ->
+          error "Trying to push a candidate that is not integrated."
```

Comment:
A big chunk of this function is now validating a condition that we know is true at the call site.
Maybe we can lift this knowledge into the type level?

Reply:
The `assert approved` will go away later once we fuse the integration status and approval status.

For the `newHead` match, I am not sure. We could accept the sha as argument and let the caller do the match, because the caller is doing the match anyway, but then there is nothing that guarantees that the sha you pass in agrees with the actual state of the pull request. (There is a similar problem with the pull request id, either we pass only that and then the function assumes that a lookup will not fail, or we pass both the id and the pull request like now, but then there is no guarantee that the two agree.)

Refinement types in Liquid Haskell could solve this, but it doesn’t seem to be as easy as turning on a language pragma.

Maybe it could be done with GADTs, with a type tag for each of the integration statuses :thinking:

So it looks like the trade-off is between trusting the caller to pass in the result of the right match, or trusting the caller to only call when the preconditions are satisfied. I have a slight preference for the latter, because for violations we can call `error`, and that is easy to debug, whereas passing in the wrong arguments could lead to silent corruption.

What do you think?

Reply:
You're right, for debugging, those checks might still be useuful.
As long as these errors are properly handled, that's fine. That's what I hinted at with my top-level comment:

> Since there are still runtime errors, would it make sense to also already address the error handling in the main function to prevent hoff from locking up again should we ever encounter those?

---

Just exploring some other ideas here:

> For the newHead match, I am not sure. We could accept the sha as argument and let the caller do the match, because the caller is doing the match anyway, but then there is nothing that guarantees that the sha you pass in agrees with the actual state of the pull request.

The only way to get a Sha at that point is extracting it from the integration status
A newtype wrapper could act as an extra safety net there making wrong code look obviously wrong:

```haskell
newtype IntegratedCommit = IntegratedCommit Sha

data IntegrationStatus = ... | Integrated IntegratedCommit BuildStatus | ...

-- Then this looks correct and typechecks
    Integrated sha buildStatus -> case buildStatus of
      BuildPending   -> pure state
      BuildSucceeded -> pushCandidate (pullRequestId, pullRequest) sha state
      BuildFailed    -> ...

-- But this looks quite wrong
    Conflicted -> pushCandidate (pullRequestId, pullRequest) (IntegratedCommit someRandomSha) state

-- And this one doesn't type check
    Conflicted -> pushCandidate (pullRequestId, pullRequest) someRandomSha state

-- And this also looks wrong if you know what the business logic is supposed to be (as in: why would one push if the build has failed:)
    Integrated sha buildStatus -> case buildStatus of
      BuildFailed  -> pushCandidate (pullRequestId, pullRequest) sha state
```

For the latter to look even more wrong, `pushCandidate` which sounds quite vague (where is it pushed to?) could be renamed to `pushCandidateToMaster`.

And lastly, `pushCandidate` is not even exposed from the module.
At some point, you'll have to draw the boundary of where you verify invariants and where you trust invariants to be upheld (although assertions checking that those assumption are correct are a good idea for debuggability).


> but then there is nothing that guarantees that the sha you pass in agrees with the actual state of the pull request

Also, this might be something where [Ghosts of Departed Proofs](https://kataskeue.com/gdp.pdf) might come in handy again.

Reply:
Okay, let’s keep it like this for now then, and re-evaluate once we fold the approval status and integration status. Maybe at that point it becomes feasible to inline this, and then the problem will go away entirely.